File: agentloom-studio/src/features/skill/api/skillApi.ts
import { apiClient } from '@/shared/api/client';
import type { PaginatedResponse } from '@/shared/types/api';
import type { ResourceSourceKind } from '@/shared/lib/resourceSource';
import type { Skill, SkillStatus } from '../types';

const BASE_PATH = 'skills';

export interface ListSkillsParams {
  page?: number;
  pageSize?: number;
  status?: SkillStatus;
  isBuiltin?: boolean;
  search?: string;
  sourceKind?: ResourceSourceKind;
}

export interface CreateSkillPayload {
  name: string;
  description?: string;
  content?: string;
  files?: File[];
}

export interface UpdateSkillPayload {
  name?: string;
  description?: string;
  content?: string;
}

function buildSearchParams(params?: ListSkillsParams): Record<string, string> {
  const searchParams: Record<string, string> = {};
  if (params?.page) searchParams.page = String(params.page);
  if (params?.pageSize) searchParams.pageSize = String(params.pageSize);
  if (params?.status) searchParams.status = params.status;
  if (params?.isBuiltin != null) searchParams.isBuiltin = String(params.isBuiltin);
  if (params?.search) searchParams.search = params.search;
  if (params?.sourceKind) searchParams.sourceKind = params.sourceKind;
  return searchParams;
}

export async function fetchSkills(
  params?: ListSkillsParams,
): Promise<PaginatedResponse<Skill>> {
  return apiClient
    .get(BASE_PATH, { searchParams: buildSearchParams(params) })
    .json<PaginatedResponse<Skill>>();
}

export async function fetchSkillById(id: string): Promise<Skill> {
  return apiClient.get(`${BASE_PATH}/${id}`).json<Skill>();
}

export async function createSkill(payload: CreateSkillPayload): Promise<Skill> {
  const { files, ...metadata } = payload;
  const formData = new FormData();
  formData.append('metadata', JSON.stringify(metadata));
  if (files) {
    for (const file of files) {
      formData.append('files', file, file.name);
    }
  }

  return apiClient
    .post(BASE_PATH, { body: formData })
    .json<Skill>();
}

export async function updateSkill(
  id: string,
  payload: UpdateSkillPayload,
): Promise<Skill> {
  const formData = new FormData();
  formData.append('metadata', JSON.stringify(payload));

  return apiClient
    .put(`${BASE_PATH}/${id}`, { body: formData })
    .json<Skill>();
}

export async function deleteSkill(id: string): Promise<void> {
  await apiClient.delete(`${BASE_PATH}/${id}`);
}

export async function archiveSkill(id: string): Promise<Skill> {
  return apiClient
    .patch(`${BASE_PATH}/${id}/archive`)
    .json<Skill>();
}

// ---------- File Management ----------

export interface SkillFileInfo {
  name: string;
  size: number;
}

export async function fetchSkillFiles(id: string): Promise<SkillFileInfo[]> {
  return apiClient.get(`${BASE_PATH}/${id}/files`).json<SkillFileInfo[]>();
}

export async function uploadSkillFile(
  id: string,
  file: File,
): Promise<SkillFileInfo> {
  const formData = new FormData();
  formData.append('file', file);
  return apiClient
    .post(`${BASE_PATH}/${id}/files`, { body: formData })
    .json<SkillFileInfo>();
}

export async function downloadSkillFile(
  id: string,
  fileName: string,
): Promise<Blob> {
  return apiClient
    .get(`${BASE_PATH}/${id}/files/${encodeURIComponent(fileName)}`)
    .blob();
}

export async function deleteSkillFile(
  id: string,
  fileName: string,
): Promise<void> {
  await apiClient.delete(
    `${BASE_PATH}/${id}/files/${encodeURIComponent(fileName)}`,
  );
}

File: agentloom_mobile/lib/features/skills/api/skill_api.dart
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../../shared/models/paginated_response.dart';
import '../../../shared/providers/api_client_provider.dart';
import '../models/skill_dto.dart';

/// 技能 API 客户端 — 对应服务端 SkillController
class SkillApi {
  final Dio _dio;

  SkillApi(this._dio);

  /// 获取技能列表（分页 + 筛选）
  Future<PaginatedResponse<SkillDto>> listSkills({
    int page = 1,
    int pageSize = 20,
    String? status,
    bool? isBuiltin,
    String? search,
  }) async {
    final queryParams = <String, dynamic>{'page': page, 'pageSize': pageSize};
    if (status != null) queryParams['status'] = status;
    if (isBuiltin != null) queryParams['isBuiltin'] = isBuiltin;
    if (search != null && search.isNotEmpty) queryParams['search'] = search;

    final response = await _dio.get(
      '/api/v1/skills',
      queryParameters: queryParams,
    );

    return PaginatedResponse<SkillDto>.fromJson(
      response.data as Map<String, dynamic>,
      (json) => SkillDto.fromJson(json! as Map<String, dynamic>),
    );
  }

  /// 获取单个技能详情
  Future<SkillDto> getSkill(String id) async {
    final response = await _dio.get('/api/v1/skills/$id');
    final body = response.data as Map<String, dynamic>;
    return SkillDto.fromJson(body['data'] as Map<String, dynamic>);
  }

  /// 创建技能
  Future<SkillDto> createSkill({
    required String name,
    String? description,
    String? content,
  }) async {
    final body = <String, dynamic>{'name': name};
    if (description != null) body['description'] = description;
    if (content != null) body['content'] = content;

    final response = await _dio.post('/api/v1/skills', data: body);
    final data = response.data as Map<String, dynamic>;
    return SkillDto.fromJson(data['data'] as Map<String, dynamic>);
  }

  /// 更新技能（含 OCC 版本）
  Future<SkillDto> updateSkill(
    String id, {
    String? name,
    String? description,
    String? content,
    required int occVersion,
  }) async {
    final body = <String, dynamic>{'occVersion': occVersion};
    if (name != null) body['name'] = name;
    if (description != null) body['description'] = description;
    if (content != null) body['content'] = content;

    final response = await _dio.put('/api/v1/skills/$id', data: body);
    final data = response.data as Map<String, dynamic>;
    return SkillDto.fromJson(data['data'] as Map<String, dynamic>);
  }

  /// 删除技能
  Future<void> deleteSkill(String id) async {
    await _dio.delete('/api/v1/skills/$id');
  }

  /// 归档技能
  Future<SkillDto> archiveSkill(String id) async {
    final response = await _dio.patch('/api/v1/skills/$id/archive');
    final data = response.data as Map<String, dynamic>;
    return SkillDto.fromJson(data['data'] as Map<String, dynamic>);
  }
}

/// 技能 API Provider
final skillApiProvider = Provider<SkillApi>((ref) {
  final dio = ref.watch(apiClientProvider);
  return SkillApi(dio);
});
